geode-core/src/integrationTest/java/org/apache/geode/internal/cache/TXDetectReadConflictJUnitTest.java

```diff
@@ -81,25 +93,95 @@ public void tearDown() throws Exception {
   public void testReadConflictsRR() throws Exception {
     cache.close();
     createCache();
-    region.put("key", "value");
-    region.put("key1", "value1");
+    region.put(key, value);
+    region.put(key1, value1);
     TXManagerImpl mgr = (TXManagerImpl) cache.getCacheTransactionManager();
     mgr.begin();
-    assertEquals("value", region.get("key"));
-    assertEquals("value1", region.get("key1"));
+    assertEquals(value, region.get(key));
+    assertEquals(value1, region.get(key1));
     mgr.commit();
   }
 
   @Test
   public void testReadConflictsPR() throws Exception {
     cache.close();
     createCachePR();
-    regionpr.put("key", "value");
-    regionpr.put("key1", "value1");
+    regionPR.put(key, value);
+    regionPR.put(key1, value1);
     TXManagerImpl mgr = (TXManagerImpl) cache.getCacheTransactionManager();
     mgr.begin();
-    assertEquals("value", regionpr.get("key"));
-    assertEquals("value1", regionpr.get("key1"));
+    assertEquals(value, regionPR.get(key));
+    assertEquals(value1, regionPR.get(key1));
     mgr.commit();
   }
+
+  @Test
+  public void readConflictsTransactionCanBlockWriteTransaction() {
+    cache.close();
+    createCache();
+
+    region.put(key, value);
+    region.put(key1, value1);
+    TXManagerImpl txManager = (TXManagerImpl) cache.getCacheTransactionManager();
+    txManager.begin();
+    assertThat(region.get(key)).isSameAs(value);
+    region.put(key1, newValue1);
+    TXState txState =
+        (TXState) ((TXStateProxyImpl) TXManagerImpl.getCurrentTXState()).getRealDeal(null, null);
+    txState.setAfterReservation(() -> readTransactionAfterReservation());
+    Runnable task = () -> doPutOnReadKeyTransaction();
+    new Thread(task).start();
```

Comment:
I would add an `ExecutorServiceRule` and use that instead. It'll ensure cleanup if the thread gets stuck for any reason and will perform error detection on it.
```
@Rule
public ExecutorServiceRule executorServiceRule = new ExecutorServiceRule();

ExecutorServiceRule.ThrowingRunnable task = () -> { Thread.sleep(100);};
executorServiceRule.submit(task);
```
I'll file a Jira to add support back in to ExecutorServiceRule for java.lang.Runnable as well. ThrowingRunnable allows the body of the run to throw anything.

Reply:
Thanks for reviewing. Will make the necessary changes.